src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1547,3 +2145,348 @@ MultiRouterPlannableQuery(Query *query, MultiExecutorType taskExecutorType,
+											newSubqueryTargetEntry);
+
+			resno++;
+		}
+	}
+
+	originalQuery->targetList = newInsertTargetlist;
+	subquery->targetList = newSubqueryTargetlist;
+
+	return NULL;
+}
+
+
+/*
+ * InsertSelectQuery returns true when the input query
+ * is INSERT INTO ... SELECT kind of query.
+ */
+bool
+InsertSelectQuery(Query *query)
+{
+	CmdType commandType = query->commandType;
+	List *rangeTableList = query->rtable;
+	RangeTblEntry *subqueryRte = NULL;
+	Query *subquery = NULL;
+
+	if (commandType != CMD_INSERT)
+	{
+		return false;
+	}
+
+	rangeTableList = query->rtable;
+	if (list_length(rangeTableList) < 2)
+	{
+		return false;
+	}
+
+	subqueryRte = lsecond(query->rtable);
+	subquery = subqueryRte->subquery;
+	if (subquery == NULL)
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * Copy a RelationRestrictionContext. Note that several subfields are copied
+ * shallowly, for lack of copyObject support.
+ */
+static RelationRestrictionContext *
+copyRelationRestrictionContext(RelationRestrictionContext *oldContext)
+{
+	RelationRestrictionContext *newContext = (RelationRestrictionContext *)
+											 palloc(sizeof(RelationRestrictionContext));
+	ListCell *relationRestrictionCell = NULL;
+
+	newContext->hasDistributedRelation = oldContext->hasDistributedRelation;
+	newContext->hasLocalRelation = oldContext->hasLocalRelation;
+	newContext->relationRestrictionList = NIL;
+
+	foreach(relationRestrictionCell, oldContext->relationRestrictionList)
+	{
+		RelationRestriction *oldRestriction =
+			(RelationRestriction *) lfirst(relationRestrictionCell);
+		RelationRestriction *newRestriction = (RelationRestriction *)
+											  palloc0(sizeof(RelationRestriction));
+
+		newRestriction->index = oldRestriction->index;
+		newRestriction->relationId = oldRestriction->relationId;
+		newRestriction->distributedRelation = oldRestriction->distributedRelation;
+		newRestriction->rte = copyObject(oldRestriction->rte);
+
+		/* can't be copied, we copy (flatly) a RelOptInfo, and then decouple baserestrictinfo */
+		newRestriction->relOptInfo = palloc(sizeof(RelOptInfo));
+		memcpy(newRestriction->relOptInfo, oldRestriction->relOptInfo,
+			   sizeof(RelOptInfo));
+		newRestriction->relOptInfo->baserestrictinfo = copyObject(
+			oldRestriction->relOptInfo->baserestrictinfo);
+
+		/* not copyable, but readonly */
+		newRestriction->plannerInfo = oldRestriction->plannerInfo;
+		newRestriction->prunedShardIntervalList = copyObject(
+			oldRestriction->prunedShardIntervalList);
+
+		newContext->relationRestrictionList =
+			lappend(newContext->relationRestrictionList, newRestriction);
+	}
+
+	return newContext;
+}
+
+
+/*
+ * ReplaceHiddenQual Replace the "hidden" partition restriction clause with
+ * the current shard's (passed in context) boundary value.
+ *
+ * Once we see ($1 = partition column), we replace it with
+ * (partCol >= shardMinValue && partCol <= shardMaxValue)
+ */
+static Node *
+ReplaceHiddenQual(Node *node, void *context)
```

Comment:
I liked `uninstantiated ` / `InstantiatePartitionQual `. I updated all references to hidden accordingly in the code/comments.